File: Source/JavaScript/engine/themeCompatibility.ts
// Copyright (c) Cratis. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

import { Theme, UiProfile } from '@cratis/scene.model';

/**
 * Finds the packages a {@link UiProfile} declares that a {@link Theme} is not declared compatible with -
 * part of Cratis/Scene#5, the same rule Screenplay's compiler already applies to warn on an incompatible
 * `ui profile`/`theme` pairing, reused here so `Scene.React` knows which packages to actually apply a
 * theme's tokens to, and so Studio's design-time tooling and Stage's build-time checks don't each
 * reimplement it. There is deliberately no implicit exemption for `core`: a theme wanting broad
 * applicability declares `compatible with core` itself, exactly like the profile/package resolver has no
 * implicit special case for it either.
 *
 * @param theme The {@link Theme} to check.
 * @param profile The {@link UiProfile} whose packages to check against.
 * @returns The packages in `profile.packages` that `theme` does not declare compatibility with.
 */
export function incompatiblePackages(theme: Theme, profile: UiProfile): string[] {
    return profile.packages.filter(package_ => !theme.compatibleWith.includes(package_));
}

/**
 * Whether a {@link Theme} is compatible with every package a {@link UiProfile} declares.
 *
 * @param theme The {@link Theme} to check.
 * @param profile The {@link UiProfile} whose packages to check against.
 * @returns `true` when `theme` declares compatibility with every package `profile` lists.
 */
export function isThemeCompatible(theme: Theme, profile: UiProfile): boolean {
    return incompatiblePackages(theme, profile).length === 0;
}

/**
 * Finds the packages a {@link Theme}'s tokens actually apply to for a given {@link UiProfile} - the
 * packages the profile activates that the theme also declares compatibility with. This is what a
 * renderer scopes token application to, rather than applying a theme's tokens globally - and is
 * recomputed on every theme switch for the live re-resolution Cratis/Scene#5 requires (no reload).
 *
 * @param theme The {@link Theme} to apply.
 * @param profile The active {@link UiProfile}.
 * @returns The packages both active in `profile` and declared compatible by `theme`.
 */
export function applicableThemePackages(theme: Theme, profile: UiProfile): string[] {
    return profile.packages.filter(package_ => theme.compatibleWith.includes(package_));
}

File: Source/DotNET/Engine/Profiles/ThemeCompatibility.cs
// Copyright (c) Cratis. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using Cratis.Scene.Model.Profiles;

namespace Cratis.Scene.Engine.Profiles;

/// <summary>
/// Checks a <see cref="Theme"/>'s <see cref="Theme.CompatibleWith"/> declaration against a
/// <see cref="UiProfile"/>'s package list - part of Cratis/Scene#5. The same rule Screenplay's compiler
/// already applies to warn on an incompatible <c>ui profile</c>/<c>theme</c> pairing, reused here so
/// Stage (build time) and Studio (design time) don't each reimplement it - and so <c>Scene.React</c> knows
/// which packages to actually apply a theme's tokens to. There is deliberately no implicit exemption for
/// <c>core</c>: a theme wanting broad applicability declares <c>compatible with core</c> itself, exactly
/// like the profile/package resolver has no implicit special case for it either.
/// </summary>
public static class ThemeCompatibility
{
    /// <summary>
    /// Finds the packages a <see cref="UiProfile"/> declares that a <see cref="Theme"/> is not declared
    /// compatible with - the same set Screenplay's compiler warns on for an incompatible pairing.
    /// </summary>
    /// <param name="theme">The <see cref="Theme"/> to check.</param>
    /// <param name="profile">The <see cref="UiProfile"/> whose packages to check against.</param>
    /// <returns>The packages in <see cref="UiProfile.Packages"/> that <paramref name="theme"/> does not declare compatibility with.</returns>
    public static IReadOnlyList<string> IncompatiblePackages(Theme theme, UiProfile profile) =>
        profile.Packages.Where(package => !theme.CompatibleWith.Contains(package)).ToList();

    /// <summary>
    /// Whether a <see cref="Theme"/> is compatible with every package a <see cref="UiProfile"/> declares.
    /// </summary>
    /// <param name="theme">The <see cref="Theme"/> to check.</param>
    /// <param name="profile">The <see cref="UiProfile"/> whose packages to check against.</param>
    /// <returns><see langword="true"/> when <paramref name="theme"/> declares compatibility with every package <paramref name="profile"/> lists.</returns>
    public static bool IsCompatible(Theme theme, UiProfile profile) => IncompatiblePackages(theme, profile).Count == 0;

    /// <summary>
    /// Finds the packages a <see cref="Theme"/>'s tokens actually apply to for a given
    /// <see cref="UiProfile"/> - the packages the profile activates that the theme also declares
    /// compatibility with. This is what a renderer scopes token application to, rather than applying a
    /// theme's tokens globally.
    /// </summary>
    /// <param name="theme">The <see cref="Theme"/> to apply.</param>
    /// <param name="profile">The active <see cref="UiProfile"/>.</param>
    /// <returns>The packages both active in <paramref name="profile"/> and declared compatible by <paramref name="theme"/>.</returns>
    public static IReadOnlyList<string> ApplicablePackages(Theme theme, UiProfile profile) =>
        profile.Packages.Where(theme.CompatibleWith.Contains).ToList();
}
